src/harness.ts

```diff
@@ -0,0 +1,493 @@
+import cssTransitions from '@dojo/widget-core/animations/cssTransitions';
+import { dom, Projection, ProjectionOptions, VNodeProperties } from 'maquette';
+import assertRender from './support/assertRender';
+import sendEvent, { SendEventOptions } from './support/sendEvent';
+
+const ROOT_CUSTOM_ELEMENT_NAME = 'test--harness';
+const WIDGET_STUB_CUSTOM_ELEMENT = 'test--widget-stub';
+const WIDGET_STUB_NAME_PROPERTY = 'data--widget-name';
+
+let harnessId = 0;
+
+const EVENT_HANDLERS = [
+	'ontouchcancel',
+	'ontouchend',
+	'ontouchmove',
+	'ontouchstart',
+	'onblur',
+	'onchange',
+	'onclick',
+	'ondblclick',
+	'onfocus',
+	'oninput',
+	'onkeydown',
+	'onkeypress',
+	'onkeyup',
+	'onload',
+	'onmousedown',
+	'onmouseenter',
+	'onmouseleave',
+	'onmousemove',
+	'onmouseout',
+	'onmouseover',
+	'onmouseup',
+	'onmousewheel',
+	'onscroll',
+	'onsubmit'
+];
+
+/**
+ * Decorate a `DNode` where any `WNode`s are replaced with stubbed widgets
+ * @param target The `DNode` to decorate with stubbed widgets
+ */
+function stubRender(target: DNode): DNode {
+	decorate(
+		target,
+		(dNode: WNode) => {
+			const { widgetConstructor, properties } = dNode;
+			dNode.widgetConstructor = StubWidget;
+			(<StubWidgetProperties> properties)._stubTag = WIDGET_STUB_CUSTOM_ELEMENT;
+			(<StubWidgetProperties> properties)._widgetName = typeof widgetConstructor === 'string'
+				? widgetConstructor
+				: (<any> widgetConstructor).name || '<Anonymous>';
+		},
+		(dNode) => isWNode(dNode)
+	);
+	return target;
+}
+
+interface StubWidgetProperties extends WidgetProperties {
+	_stubTag: string;
+	_widgetName: string;
+}
+
+class StubWidget extends WidgetBase<StubWidgetProperties> {
+	render(): DNode {
+		const { bind, _stubTag: tag, _widgetName: widgetName } = this.properties;
+		return v(tag, { bind, [WIDGET_STUB_NAME_PROPERTY]: widgetName }, this.children);
+	}
+}
+
+interface SpyRenderMixin {
+	spyRender(result: DNode): DNode;
+}
+
+interface SpyTarget {
+	actualRender(actual: DNode): void;
+}
+
+/**
+ * A mixin that adds a spy to the render process
+ * @param base The base class to add the render spy to
+ * @param target An object with a property named `lastRender` which will be set to the result of the `render()` method
+ */
+function SpyRenderMixin<T extends Constructor<WidgetBase<WidgetProperties>>>(base: T, target: SpyTarget): T & Constructor<SpyRenderMixin> {
+
+	class SpyRender extends base {
+		@afterRender()
+		spyRender(result: DNode): DNode {
+			target.actualRender(result);
+			return stubRender(result);
+		}
+	};
+
+	return SpyRender;
+}
+
+/**
+ * A private class that is used to actually render the widget and keep track of the last render by
+ * the harnessed widget.
+ */
+class WidgetHarness<P extends WidgetProperties, W extends typeof WidgetBase> extends WidgetBase<P> {
+	private _widgetConstructor: W;
+	private _afterCreate: (element: HTMLElement) => void;
+	private _id: string = ROOT_CUSTOM_ELEMENT_NAME + '-' + (++harnessId);
```

Comment:
Hmmm... for some reason, the trivially inferred type was not flagged.  I wonder if `tslint` didn't detect the implicit coercion, though TypeScript does.  Anyways, changed.